### DISCUSSION TITLE
Is it possible to check if an error was raised?

### DISCUSSION THREAD

**Discussion by stephannv** [MikeSchulze/gdUnit4#239] [Q&A] [answered]
For example, on RSpec:
```ruby
def div(a, b)
  raise "Division by 0 error" if b == 0

  a / b
end

describe "#div" do
  context "when b is zero" do
    it "raises error" do
      calculator = Calculator.new

      expect { calculator.div(3, 0) }.to raise_error(RuntimeError)
      # or
      expect { calculator.div(3, 0) }.to raise_error("Division by 0 error")
      # or
      expect { calculator.div(3, 0) }.to raise_error
    end
  end
end
```

On GDScript I have this class:
```gdscript
class_name Grid

func _init(width: int, height: int) -> void:
  if width % 2 == 0 or height % 2 == 0: assert(false, "Grid must be odd-sized")
  # or
  if width % 2 == 0 or height % 2 == 0: push_error("Grid must be odd-sized")
```

Is it possible to test this scenario?

**Comment by MikeSchulze** [comment]
hi @stephannv yes for this scenario you can enable the error to failure reporting in the GdUnit settings.
![image](https://github.com/MikeSchulze/gdUnit4/assets/347037/5155f6a7-2bc1-4027-b90b-f7837db86bcd)

[Report Settings](https://mikeschulze.github.io/gdUnit4/first_steps/settings/#report-settings)
* Godot
  * Push Error
  * Script Error
  
  
 It will catch both cases
 
![image](https://github.com/MikeSchulze/gdUnit4/assets/347037/50bb4793-6ea7-476a-9ce3-e34747bc3b25)
![image](https://github.com/MikeSchulze/gdUnit4/assets/347037/8fddf9e5-1b63-4864-a8ff-5b1c9a178089)

**Reply by stephannv** [reply]
@MikeSchulze but will this make test fail? My idea is test be successful, eg.:
```gdscript
assert_lambda(func(): 
  Grid.new(12, 12)
).raise_error("Grid must be odd-sized")
```

**Reply by MikeSchulze** [reply]
I do not know this function `assert_lambda`, who offers this function?

**Reply by stephannv** [reply]
@MikeSchulze it is just an exemple of what I'm looking for. I think it doesn't exist. I'm looking for a way to assert if error was raised without failing tests

**Reply by MikeSchulze** [reply]
you can use `assert_failure( <lambda>)`
```gdscript
assert_failure(func(): assert_str("This is a test message").is_equal("This is a test Message")) \
		.is_failed() \
		.has_message("Expecting:\n 'This is a test Message'\n but was\n 'This is a test Mmessage'")
````

feel free to join the discord server asking for help

**Reply by stephannv** [reply]
@MikeSchulze Does it work only if there is an assertion inside proc/lambda? Can I put other things inside? Example:
```gdscript
assert_failure(func(): 
  grid = Grid.new()
  grid.do_something() # this methods raises an error
).is_failed().has_message("Some error message")
```

**Comment by MikeSchulze** [comment, marked as answer]
Ok, if I understand you correctly, you want to test if a certain call generates an error and test for it.
```
class MyClass:
	
	func foo(value :int):
		match value:
			1: assert(false, "foo failed")
			2: push_error("foo failed")

func test_foo():
	assert_error(func(): 
		MyClass.new().foo(0) # success
	).is_success()

	assert_error(func(): 
		MyClass.new().foo(1) # fails by an assert
	).is_failed().has_message("foo failed")
	
	assert_error(func(): 
		MyClass.new().foo(2) # fails by an push_error
	).is_failed().has_message("foo failed")
```

sorry, for this kind of scenario no asserts exists at the moment, feel free to create a feature request

**Reply by stephannv** [reply]
Yes, exactly this I was looking for. 

Thanks @MikeSchulze for the support.

**Comment by MikeSchulze** [comment]
@stephannv the next version v4.1.4 will provide the new `assert_error`
Than you can test for runtime errors

```gdscript
func test_is_success() -> void:
	await assert_error(func (): GodotErrorTestClass.new().test(0)).is_success()


func test_is_assert_failed() -> void:
	await assert_error(func (): GodotErrorTestClass.new().test(1))\
		.is_runtime_error('Assertion failed: this is an assert error')



func test_is_push_warning() -> void:
	await assert_error(func (): GodotErrorTestClass.new().test(2))\
		.is_push_warning('this is an push_warning')


func test_is_push_error() -> void:
	await assert_error(func (): GodotErrorTestClass.new().test(3))\
		.is_push_error('this is an push_error')


func test_is_runtime_error() -> void:
	await assert_error(func (): GodotErrorTestClass.new().test(4))\
		.is_runtime_error("Division by zero error in operator '/'.")
```

**Reply by stephannv** [reply]
Nice @MikeSchulze. LGTM
